File: backend/node/api/src/application/use-cases/get-pipeline.ts
import { inject } from "../../infra/DI/container";
import type { PipelineRepository } from "../../infra/repository/pipeline";

type Input = string;

type Output = {
	id: string;
	name?: string;
	description?: string;
	initialStepId?: string;
	createdAt: Date;
	updatedAt: Date;
};

export class GetPipelineUseCase {
	@inject("pipeline-repository")
	private declare readonly pipelineRepository: PipelineRepository;

	async execute(
		input: Input,
	): Promise<[Output, undefined] | [undefined, Error]> {
		const [pipeline, error] = await this.pipelineRepository.getById(input);
		if (error) return [undefined, error];

		const output: Output = {
			id: pipeline.getId(),
			name: pipeline.getName(),
			description: pipeline.getDescription(),
			initialStepId: pipeline.getInitialStepId(),
			createdAt: pipeline.getCreatedAt(),
			updatedAt: pipeline.getUpdatedAt(),
		};

		return [output, undefined];
	}
}

File: backend/go/api/internal/pipeline/use-cases/get-pipeline.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/pipeline"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/shared/errors"
)

type GetPipelineUseCaseDependencies struct {
	PipelineRepository *pipeline.PipelineRepository
}

type GetPipelineUseCase struct {
	GetPipelineUseCaseDependencies
}

func NewGetPipelineUseCase(deps GetPipelineUseCaseDependencies) GetPipelineUseCase {
	return GetPipelineUseCase{
		GetPipelineUseCaseDependencies: deps,
	}
}

type GetPipelineInput struct {
	ID string
}

type GetPipelineOutput struct {
	ID            string
	Name          *string
	Description   *string
	InitialStepId *string
	Canvas        string
	CreatedAt     string
	UpdatedAt     string
}

func (u *GetPipelineUseCase) Execute(input GetPipelineInput) (GetPipelineOutput, error) {
	p, err := u.PipelineRepository.GetByID(context.Background(), input.ID)
	if err != nil {
		return GetPipelineOutput{}, err
	}

	if p == nil {
		return GetPipelineOutput{}, errors.PipelineNotFoundError(input.ID)
	}

	return GetPipelineOutput{
		ID:            p.GetId(),
		Name:          p.GetName(),
		Description:   p.GetDescription(),
		InitialStepId: p.GetInitialStepId(),
		Canvas:        p.GetCanvas(),
		CreatedAt:     p.GetCreatedAt(),
		UpdatedAt:     p.GetUpdatedAt(),
	}, nil
}
